### ISSUE TITLE
llvmlite override expects missing file

### ISSUE THREAD

**Issue by colehaus** [nix-community/poetry2nix#1115]
When trying to build a project, I get:

```
error: builder for '/nix/store/ifpp1dsybn76sa0p0w98zfac1nrim0ms-python3.10-llvmlite-0.39.1.drv' failed with exit code 1;
       last 10 log lines:
       > Using pythonImportsCheckPhase
       > Sourcing python-namespaces-hook
       > Sourcing python-catch-conflicts-hook.sh
       > Sourcing wheel setup hook
       > unpacking sources
       > Executing wheelUnpackPhase
       > Finished executing wheelUnpackPhase
       > patching sources
       > substitute(): ERROR: file 'ffi/Makefile.linux' does not exist
       > /nix/store/5s1yg5l36wzgy1dj0vv1ibarc4g7vrdr-stdenv-linux/setup: line 136: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/ifpp1dsybn76sa0p0w98zfac1nrim0ms-python3.10-llvmlite-0.39.1.drv'.
```

I believe this is due to the attempted override here:

https://github.com/nix-community/poetry2nix/blob/master/overrides/default.nix#L1115

I presume it's because the llvmlite version I'm attempting (0.39.1) is newer than anything included and expected in the override file.

I'm able to work around the issue by adding the override:

```nix
            llvmlite = python.pkgs.llvmlite;
```

**Comment by SenchoPens** [comment]
IMHO I don't think it's a poetry2nix-specific bug, and I suspect it might've happened because you had `preferWheels` turned on (which can lead to strange behaviour as mentioned in some other issue).

I had a similar error when I tried to install `umap-learn`, which depends on `numba`, which in turn depends on `llvmlite`.
As I didn't want to disable `preferWheels` for all packages, I've used this code in my poetry overrides:

`nix
let disableWheel = package: package.override {
  preferWheel = false;
  enableParallelBuilding = true;   # optional
};
in poetryOverrides.withDefaults
  (self: super: lib.genAttrs ["umap-learn" "llvmlite" "numba"] (name: disableWheel super.${name}));
`

FYI the code above built, but later in Python `numba` failed to import because `llvmlite` was built with legacy LLVM version. If you encounter the same issue, I've just created a PR:  #1164

---

**TL;DR** Check if you have set `preferWheels = true`, and turn this off either for all packages, or for the `llvmlite` and for the packages that depend on it that you're installing.

**Comment by knkski** [comment]
I could not get `preferWheel = true;` to work for me, until I realized that I had a line like this:

```nix
overrides = poetry2nix.defaultPoetryOverrides.extend (<...>);
```

Which worked for overriding some things, but not `preferWheel`. After I switched to this override style, `preferWheel` worked nicely for me:

```nix
overrides = poetry2nix.overrides.withDefaults (<...>);
```

I'm not really sure what the difference is, or how I ended up with the prior option in the first place. For what it's worth, here's my full `shell.nix` file that allowed me to install `shap` properly:

```nix
with import <nixpkgs> { };
let
  poetry2nix = import (builtins.fetchGit {
    url = "https://github.com/nix-community/poetry2nix";
    rev = "4f91d45e39cf64b642c6d2725c7bd50d6fcb544b";
  }) { pkgs = pkgs; };

  deps = poetry2nix.mkPoetryEnv {
    python = pkgs.python310;
    projectDir = ./.;
    preferWheels = true;
    groups = [ "dev" ];
    overrides = poetry2nix.overrides.withDefaults (self: super: {
      llvmlite = super.llvmlite.override { preferWheel = false; };
      numba = super.numba.overridePythonAttrs (old: {
        buildInputs = (old.buildInputs or [ ]) ++ [ pkgs.tbb_2021_8 ];
      });
    });
  };

in pkgs.mkShell rec {
  buildInputs = [ pkgs.zlib ];
  packages = [ deps pkgs.poetry pkgs.nixfmt ];
  shellHook = ''
    export LD_LIBRARY_PATH="${
      pkgs.lib.makeLibraryPath buildInputs
    }:$LD_LIBRARY_PATH"
    export LD_LIBRARY_PATH="${pkgs.stdenv.cc.cc.lib.outPath}/lib:$LD_LIBRARY_PATH"
  '';
}
```
